File: backend/app/schemas/trace.py
from __future__ import annotations
from typing import Literal, Optional
from pydantic import BaseModel


class RegisterState(BaseModel):
    rax: str; rbx: str; rcx: str; rdx: str
    rsi: str; rdi: str; rsp: str; rbp: str
    rip: str; rflags: str


InstructionCategory = Literal[
    "data_move", "arithmetic", "control_flow", "syscall", "memory", "crypto"
]
BehaviorCategory = Literal[
    "filesystem", "network", "memory", "privilege", "process", "crypto"
]


class MemoryAccess(BaseModel):
    address: str
    size: int
    type: Literal["read", "write"]
    value: Optional[str] = None


class InstructionStep(BaseModel):
    index: int
    address: str
    bytes: str
    mnemonic: str
    operands: str
    category: InstructionCategory
    isSyscall: bool
    syscallName: Optional[str] = None
    registers: RegisterState
    memoryAccess: Optional[MemoryAccess] = None
    annotation: str


class SyscallEvent(BaseModel):
    index: int
    name: str
    number: int
    args: list[str]
    returnValue: str
    category: BehaviorCategory


class TraceData(BaseModel):
    programName: str
    architecture: Literal["x86_64"]
    totalInstructions: int
    steps: list[InstructionStep]
    syscalls: list[SyscallEvent]

File: frontend/src/types/trace.ts
export interface RegisterState {
  rax: string
  rbx: string
  rcx: string
  rdx: string
  rsi: string
  rdi: string
  rsp: string
  rbp: string
  rip: string
  rflags: string
}

export type InstructionCategory =
  | 'data_move'
  | 'arithmetic'
  | 'control_flow'
  | 'syscall'
  | 'memory'
  | 'crypto'

export type BehaviorCategory =
  | 'filesystem'
  | 'network'
  | 'memory'
  | 'privilege'
  | 'process'
  | 'crypto'

export interface MemoryAccess {
  address: string
  size: number
  type: 'read' | 'write'
  value?: string
}

export interface InstructionStep {
  index: number
  address: string
  bytes: string
  mnemonic: string
  operands: string
  category: InstructionCategory
  isSyscall: boolean
  syscallName?: string
  registers: RegisterState
  memoryAccess?: MemoryAccess
  annotation: string
}

export interface SyscallEvent {
  index: number
  name: string
  number: number
  args: string[]
  returnValue: string
  category: BehaviorCategory
}

export interface TraceData {
  programName: string
  architecture: 'x86_64'
  totalInstructions: number
  steps: InstructionStep[]
  syscalls: SyscallEvent[]
}
